src/wallet/rpcwallet.cpp

```diff
@@ -895,38 +905,35 @@ static UniValue getbalance(const JSONRPCRequest& request)
 
     LOCK2(cs_main, pwallet->cs_wallet);
 
-    if (IsDeprecatedRPCEnabled("accounts")) {
-        const UniValue& account_value = request.params[0];
-        const UniValue& minconf = request.params[1];
-        const UniValue& include_watchonly = request.params[2];
+    const UniValue& account_value = request.params[0];
 
-        if (account_value.isNull()) {
-            if (!minconf.isNull()) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER,
-                    "getbalance minconf option is only currently supported if an account is specified");
-            }
-            if (!include_watchonly.isNull()) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER,
-                    "getbalance include_watchonly option is only currently supported if an account is specified");
-            }
-            return ValueFromAmount(pwallet->GetBalance());
-        }
+    int min_depth = 0;
+    if (IsDeprecatedRPCEnabled("accounts") && !account_value.isNull()) {
+        // Default min_depth to 1 when an account is provided.
+        min_depth = 1;
+    }
+    if (!request.params[1].isNull()) {
+        min_depth = request.params[1].get_int();
+    }
+
+    isminefilter filter = ISMINE_SPENDABLE;
+    if (!request.params[2].isNull() && request.params[2].get_bool()) {
+        filter = filter | ISMINE_WATCH_ONLY;
+    }
+
+    if (!account_value.isNull()) {
 
         const std::string& account_param = account_value.get_str();
         const std::string* account = account_param != "*" ? &account_param : nullptr;
 
-        int nMinDepth = 1;
-        if (!minconf.isNull())
-            nMinDepth = minconf.get_int();
-        isminefilter filter = ISMINE_SPENDABLE;
-        if(!include_watchonly.isNull())
-            if(include_watchonly.get_bool())
-                filter = filter | ISMINE_WATCH_ONLY;
-
-        return ValueFromAmount(pwallet->GetLegacyBalance(filter, nMinDepth, account));
+        if (!IsDeprecatedRPCEnabled("accounts") && *account != "") {
```

Comment:
In commit "[RPC] [wallet] allow getbalance to use min_conf and watch_only"

I think the safest thing here would be to raise an error whenever accounts are deprecated and account is non-null, because previously `getbalance(null)`, `getbalance("*")`, and `getbalance("")` each did different things, and it doesn't seem good to silently ignore a requested behavior. Another slightly less safe but maybe acceptable option would be to change `""` to `"*"` on this line, since `getbalance("*")` and `getbalance(null)` at least were more similar than `getbalance("*")` and `getbalance("")`

Previously, and currently with the deprecated accounts feature enabled:

- `getbalance(null)` would return total balance
- `getbalance("*")` would return total balance with legacy accounting method
- `getbalance("account")` would return balance associated with a named account
- `getbalance("")` would return leftover balance not associated with any named accounts

Reply:
https://github.com/bitcoin/bitcoin/pull/13566#discussion_r199161775

Also want to note that another approach we could take here would not be to deprecate the `account` argument at all, but instead just rename it to `label` and pass it down as an option to the GetBalance() and GetAvailableCredit methods where it could be used to filter txouts there by address label. This would be a simple code change, just adding `if (!label || *label == GetLabelName(txout.scriptPubKey))`.